tests/integration/test_charm_with_loki.py

```diff
@@ -0,0 +1,210 @@
+    await ops_test.model.wait_for_idle(apps=[loki_tester_app_name], status="active")
+    await asyncio.gather(
+        *(
+            assert_logs_in_loki(mode, loki_app_name, ops_test, loki_address)
+            for mode in modes.split(",")
+        )
+    )
+
+
+stream_template = {
+    "job": "juju_{model_name}_{uuid}_loki-tester_loki-tester_0_loki-tester",
+    "juju_application": "loki-tester",
+    "juju_charm": "loki-tester",
+    "juju_model": "{model_name}",
+    "juju_model_uuid": "{uuid}",
+    "juju_unit": "loki-tester/0",
+}
+
+
+async def assert_logs_in_loki(mode: str, loki_app_name: str, ops_test: OpsTest, loki_address: str):
+    url = f"http://{loki_address}:3100"
+
+    async def query_job(job_name, attempt=0):
+        params = {"query": '{job="%s"}' % job_name}
+        # query_range goes from now to up to 1h ago, more
+        # certain to capture something
+        query_url = f"{url}/loki/api/v1/query_range"
+        jsn = requests.get(query_url, params=params).json()
+        results = jsn["data"]["result"]
+        if not results:
+            raise ValueError(f"{query_url} returned an empty result set.")
+        result = results[0]
+        return result
+
+    if mode == "loki":
+        res = await query_job(TEST_JOB_NAME)
+        assert res["stream"] == {"job": TEST_JOB_NAME}
+        assert res["values"][0][1] == LOKI_LOG_MSG
+        return
+
+    if mode == "file":
+        suffix = "_loki-tester"
+        msg = FILE_LOG_MSG
+        extra_stream = {"filename": "/loki_tester_msgs.txt"}
+        job_suffix = ""
+    else:  # syslog mode
+        suffix = "_loki-tester_syslog"
+        # need to prefix BOM \ufeff because syslog msgs are encoded in some
+        # special way. 'utf-8-sig' perhaps?
+        msg = "\ufeff" + SYSLOG_LOG_MSG
+        extra_stream = {}
+        job_suffix = "_syslog"
+
+    # get the job name for the log stream
+    labels = requests.get(f"{url}/loki/api/v1/label/job/values").json().get("data")
+    if not labels:
+        raise RuntimeError("no jobs present: nothing was logged yet.")
+    try:
+        job_label = next(filter(lambda x: x.endswith(suffix), labels))
+    except StopIteration:
+        raise RuntimeError(f"expected label with suffix {suffix} not " f"found in" f"{labels}.")
+
+    print("job label:", job_label)
+
+    res = await query_job(job_label)
+    juju_model_uuid_key = "juju_model_uuid"
+    juju_model_uuid = res["stream"][juju_model_uuid_key]
+    expected_stream = stream_template.copy()
+    expected_stream.update(extra_stream)
+    expected_stream["juju_model"] = juju_model_name = res["stream"]["juju_model"]
+    expected_stream[juju_model_uuid_key] = juju_model_uuid
+    expected_stream["job"] = (
+        expected_stream["job"].format(uuid=juju_model_uuid, model_name=juju_model_name)
+        + job_suffix
+    )
+
+    assert res["stream"] == expected_stream
+    assert res["values"][0][1] == msg
```

Comment:
the one checking that `msg` matches is essential, because we could be mixing up the streams and be receiving logs from syslog when we are in fact checking that the file route is working.
The other one feels more skippable, although I thought it'd be useful to verify that the tags are correctly being applied. 
Maybe that is not something for an itest, however. I can remove that if it improves clarity.

Reply:
Ok, I split up the test and rethought the flow. I think it is easier to follow now, let me know what you think.